url90b6689baa_adsonrocha_PD321_tgz-pJ8-MyFuzzyLiteClassJ8/src/com/fuzzylite/Engine.java

```diff
@@ -0,0 +1,643 @@
+            message.append("- Engine has no rule blocks\n");
+        }
+        for (int i = 0; i < this.ruleBlocks.size(); ++i) {
+            RuleBlock ruleBlock = this.ruleBlocks.get(i);
+            if (ruleBlock == null) {
+                message.append(String.format(
+                        "- Engine has a null rule block at index <%d>\n", i));
+            } else {
+                if (ruleBlock.getRules().isEmpty()) {
+                    message.append(String.format(
+                            "- Rule block <%s> has no rules\n", ruleBlock.getName()));
+                }
+                int requiresConjunction = 0;
+                int requiresDisjunction = 0;
+                for (int r = 0; r < this.ruleBlocks.size(); ++r) {
+                    Rule rule = ruleBlock.getRule(r);
+                    if (rule == null) {
+                        message.append(String.format(
+                                "- Rule block <%s> has a null rule at index <%d>\n",
+                                ruleBlock.getName(), r));
+                    } else {
+                        int thenIndex = rule.getText().indexOf(" " + Rule.FL_THEN + " ");
+                        if (rule.getText().indexOf(" " + Rule.FL_AND + " ") < thenIndex) {
+                            ++requiresConjunction;
+                        }
+                        if (rule.getText().indexOf(" " + Rule.FL_OR + " ") < thenIndex) {
+                            ++requiresDisjunction;
+                        }
+                    }
+                }
+                if (requiresConjunction > 0 && ruleBlock.getConjunction() == null) {
+                    message.append(String.format(
+                            "- Rule block <%s> has no Conjunction\n", ruleBlock.getName()));
+                    message.append(String.format(
+                            "- Rule block <%s> has %d rules that require Conjunction", ruleBlock.getName(), requiresConjunction));
+                }
+                if (requiresDisjunction > 0 && ruleBlock.getDisjunction() == null) {
+                    message.append(String.format(
+                            "- Rule block <%s> has no Disjunction\n", ruleBlock.getName()));
+                    message.append(String.format(
+                            "- Rule block <%s> has %d rules that require Disjunction", ruleBlock.getName(), requiresDisjunction));
+                }
+                if (ruleBlock.getActivation() == null) {
+                    message.append(String.format(
+                            "- Rule block <%s> has no Activation\n", ruleBlock.getName()));
+                }
+            }
+        }
+        return message.length() == 0;
+    }
+
+    @Override
+    public String toString() {
+        return new FclExporter().toString(this);
+    }
+
+    public enum Type {
+
+        NONE, MAMDANI, LARSEN, TAKAGI_SUGENO, TSUKAMOTO, INVERSE_TSUKAMOTO, UNKNOWN;
+    };
+
+    public Type type() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 26. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 306-329, lines 332-346, lines 348-363, lines 365-378 into separate methods.